src/make.ts

```diff
@@ -95,15 +95,58 @@ function error<T>(error: string): Make<T> {
   return Make.error<T>([{ path: [], error }]);
 }
 
-function checkString(value: any) {
-  if (typeof value !== 'string') {
-    return getErrorWithValueMsg('expected a string', value);
+export function registerFormat(format: string, maker: Maker<any, undefined>) {
+  assert(!formats[format], `format ${format} is already registered`);
+  formats[format] = maker;
+}
+
+const formats: Record<string, Maker<any, undefined>> = {};
+
+function getFormatter(format: string | undefined): Maker<any, undefined> {
+  if (format === undefined) {
+    return () => Make.ok(undefined);
   }
-  return Make.ok(value);
+  const formatter = formats[format];
+  if (!formatter) {
+    return () => Make.ok(undefined);
+  }
+  return formatter;
+}
+
+function getPatterner(pattern: string | undefined): Maker<any, void> {
+  if (pattern === undefined) {
+    return value => Make.ok(value);
+  }
+  const reg = new RegExp(pattern);
```

Comment:
This will throw if the regex in invalid (tried with ` new RegExp('([a-z')`). Should we perhaps catch it and re-throw it a nice message?

Reply:
there is even a test that this throws

Reply:
oh, my bad. haven't looked at the tests yet